runtime/Cpp/runtime/src/misc/IntervalSet.h

```diff
@@ -30,15 +31,17 @@ namespace misc {
   protected:
     /// The list of sorted, disjoint intervals.
     std::vector<Interval> _intervals;
-    bool _readonly;
+    std::atomic<bool> _readonly;
```

Comment:
The idea here is that you can check _readonly orders of magnitude more quickly than taking a lock on a mutex. That's pretty fast, and in the scheme of all the other stuff going on (eg. copying lots of IntervalSet) not too bad.

For optimising IntervalSet, adding `IntervalSet(IntervalSet&&)` and `IntervalSet& operator=(IntervalSet&&)` and using `std::move` is likely to be a big win. The other question is whether the copy constructor should really call `addAll()` at all because the source instance will have already done all the checking in `add(Interval const&)`. 

IntervalSets seem to be copied a lot in the code, this is likely to improve things.